### ISSUE TITLE
Create a installation script

### ISSUE THREAD

**Issue by NuSkooler** [NuSkooler/enigma-bbs#44]
It would be nice to have a installation script that users -- especially those completely new to Node.js -- can use to quickly get up and running.

Example:
```
curl -o- https://github.com/NuSkooler/enigma-bbs/misc/install.sh | bash
```
(Note: **Just an example!** The above does not currently exist!)

The script should try to work on most popular *nix environments. Windows will require something else...

**Comment by NuSkooler** [comment]
Probably provide two paths/choices:
  1. With [nvm](https://github.com/creationix/nvm)
  2. Standard Node.js installation

Should **never** install requiring root

**Comment by tracker1** [comment]
For windows, you could setup build/release that could zip up the entire thing, after npm install, with a copy of node.exe in the project directory... 

Alternative might be to use electron as a base, and then could integrate UI features, since most windows users are likely to run this under the context of a specific user.

Additional reference:
[1] https://github.com/stephenwvickers/node-os-service

**Comment by NuSkooler** [comment]
I think for now this should be a *nix task with a separate Windows task. Reason being that the source should be distributed in all cases with the ability to git pull or npm update... which of course is a issue if binaries are distributed in order to not require Visual Studio or the like. 

Perhaps when we're at the 1.0 point (e.g. out of alpha/beta) releases could be provided in more packaged up forms.

**Comment by NuSkooler** [comment]
> [1] https://github.com/stephenwvickers/node-os-service

This is looks pretty cool for Windows. I'm currently running Xibalba via [PM2](https://github.com/Unitech/pm2) on CentOS.

**Comment by ellisgl** [comment]
If you don't need load balancing and your version of Linux runs Systemd, you can just create a Systemd profile for it.

**Comment by NuSkooler** [comment]
> If you don't need load balancing and your version of Linux runs Systemd, you can just create a Systemd profile for it.

Yeah I like having some of the additional pm2 functionality. This kinda stuff does seem like it should be put in the docs somewhere.

**Comment by davestephens** [comment]
I'm gonna have a stab at this, I was thinking about doing it anyway just so I can reinstall if/when my Pi dies...

**Comment by NuSkooler** [comment]
@davestephens FWIW, I'd like to stick with installing Node itself via `nvm` as installing Node.js via package management and the like should be banned from the earth (installs stuff as root/etc.). This would ensure the entire process does not need root/sudo and Node versions can be switched easily.

With that said, if people feel strongly about it, I think the script could detect existing node installations and prompt to use them, but default to the nvm approach.

Hopefully that makes sense :smiling_imp:

**Comment by apamment** [comment]
Should it really be installing node.js at all? I mean, wouldn't that be like a website that installs apache? Just a thought.

**Comment by tracker1** [comment]
Maybe do a `node --version` test, if no node, then prompt the user that they need python 2.7.x, dev tools, and node 4.x installed as prerequisites (build tools and python are required for sqlite3 module).

**Comment by NuSkooler** [comment]
@apamment : The intent of this script is for 100% newbies that don't much know/care about Node.js (there have been quite a few) such that they could simply run the command and out pops a working system. With that said, I think Node.js installation is required.

However, it would certainly be nice to prompt them for such things.

@tracker1 : Python is not required on *nix systems (which this script is targeted for). Windows is a whole different mess. As stated in a couple comments above, it would be good to check for existing Node.js installs and prompt rather to use them or install a new (with the recommended version).

**Comment by apamment** [comment]
Yep fair enough about the node.js install.

Python is required for *nix - if it needs to build certain modules, python just happens to be installed most of the time. I discovered this on my NetBSD install (python was installed as python2 or something and npm bailed because it couldn't find python - even with the python environment variable)

**Comment by davestephens** [comment]
I guess we need to decide how far the script should go with helping out a
newbie getting their system up to scratch. All *popular* *nix distributions
will have Python installed by default, as well as all the usual bins you
need like grep, which, etc which are "just there". I'd argue that if you
were tinkering with BSD then you're a bit beyond a newbie and just saying
"hey, you're missing abc package" and exiting is enough.

So here's an overview:

* Check Python, git, curl etc, exit with a hint to fix if not available
* Check for / install nvm
* Install Node 4.4.x
* Get Enigma
* nvm install
* Run config generator

...anything obvious I'm missing here?

**Comment by NuSkooler** [comment]
@davestephens I agree with all you said.

**Comment by apamment** [comment]
Sounds good. Raspberry Pi users may have trouble with NVM, but that could be just addressed with an error message.

**Comment by NuSkooler** [comment]
@apamment RPi is so popular with BBS users, I do think we should address it in some way. At some point it may be worth detecting RPi and doing an alternate installation.

**Comment by davestephens** [comment]
I'm using nvm on a RPi1 and it works fine, what issue(s) were you expecting?

It took a while to run through compiling the native extensions, but is
absolutely fine aside from that...

**Comment by apamment** [comment]
Hmm that's odd. Perhaps they fixed it? I had errors saying it didn't support building - now that I think about it it might have been my odroid.. (too many little computers lol)

**Comment by NuSkooler** [comment]
@davestephens What @apamment said -- previously they didn't have compatible arm builds and building had been deprecated. I don't know which they fixed, but this is good news!

**Comment by tracker1** [comment]
There's been a lot of effort to get better platform build-support for ARM since the iojs re-merge...

**Comment by davestephens** [comment]
Alriiiiiiiiiiight, so I'm pretty much there with this for a first stab. In my testing I hit the sqlite-won't-compile issue once, but I can't seem to reproduce it despite winding my system back to a clean state. Further testing would be appreciated - there are bound to be some distribution specific things I haven't seen that I could iron out with more info.

The command to run it is slightly different to Bryan's example, primarily because you can't pipe into bash _and_ accept user input - which is required to run the config generator. So, the command to launch it becomes:

````bash < (curl -o- <url_to_install_script>)````

Feel free to have a nose at https://github.com/NuSkooler/enigma-bbs/pull/87. 

Thoughts appreciated!

**Comment by NuSkooler** [comment]
@davestephens Various comments added to #87. I'd like your thoughts on some of them, maybe a couple tweaks, and I think we're in business! Thanks for your efforts!

**Comment by NuSkooler** [comment]
Merged in @davestephens script from #87

Thanks again for the contribution - this should help newcomers a lot!

Reminder to self: Need to update docs
